ipaserver/plugins/servicedelegation.py

```diff
@@ -212,22 +223,39 @@ def post_callback(self, ldap, completed, failed, dn, entry_attrs,
         failed[self.principal_failedattr] = {}
         failed[self.principal_failedattr][self.principal_attr] = []
         names = options.get(self.member_names[self.principal_attr], [])
-        ldap_obj = self.api.Object['service']
+        basedn = self.api.env.container_accounts + self.api.env.basedn
         if names:
             for name in names:
                 if not name:
                     continue
-                name = normalize_principal(name)
-                obj_dn = ldap_obj.get_dn(name)
+
                 try:
-                    ldap.get_entry(obj_dn, ['krbprincipalname'])
+                    n = kerberos.Principal(name, realm=self.api.env.realm)
+                except ValueError:
+                    raise errors.ValidationError(
+                        name='principal',
+                        reason=_("Malformed principal: %(error)s") % dict(
+                            error=str(e)))
+
+                if len(n.components) == 1 and (n.components[0][-1] != '$'):
+                    nprincipal = 'host/' + unicode(n)
+                else:
+                    nprincipal = unicode(n)
+                try:
+                    e_attrs = ldap.find_entry_by_attr(
+                        'krbprincipalname', nprincipal, 'krbprincipalaux',
+                        attrs_list=['krbprincipalname'],
+                        base_dn=basedn)
                 except errors.NotFound as e:
                     failed[self.principal_failedattr][
                         self.principal_attr].append((name, unicode(e)))
                     continue
                 try:
-                    if name not in entry_attrs.get(self.principal_attr, []):
-                        members.append(name)
+                    mprinc = [p for p in e_attrs.get('krbprincipalname')
+                              if nprincipal.lower() == unicode(p).lower()]
+                    mprinc = unicode(mprinc.pop())
```

Comment:
This block is hard to understand. It looks like you are looking for the last entry (or any entry?) that has matches the principal in lower case. A for loop would be easier to understand:

```
for p in e_attrs.get('krbprincipalname'):
    p = unicode(p)
    if p.lower() == nprincipal.lower():
        mprinc = p
        break
```